### DISCUSSION TITLE
New MQTT topic for Location combining Latitude and Longitude

### DISCUSSION THREAD

**Discussion by jlpouffier** [teslamate-org/teslamate#3058] [Ideas]
**Context**
I have TeslaMate and I integrated the most important values with Home Assistant using MQTT.
One of my integration is the location of the car, so that I can do crazy stuff like heating my place when I am close to home, opening / closing the garage door, etc, etc.
(I love it FYI - Amazing project)

**Problem**
Since both Latitude and longitude comes on different MQTT topics:
- teslamate/cars/$car_id/latitude
- teslamate/cars/$car_id/longitude

Then even if my Home Assistant instance hosting my MQTT broker is powerful, both message will **never** be processed at the same time. One will **always** be processed before the other. (This is the definition of the Queue in M**Q**TT 😊)

This leads to the following issue (I agree with you, it's a small _issue_, but still):

on TeslaMate I see the different location points making perfectly accurate driving lines:
![Screenshot 2023-01-13 at 15 02 36](https://user-images.githubusercontent.com/5878296/212338461-dcdba873-2914-47f8-aa16-e6eea13dde48.png)

Whereas on Home Assistant, I see this:
![Screenshot 2023-01-13 at 15 02 29](https://user-images.githubusercontent.com/5878296/212338508-342eb709-bfe6-4e54-9c96-5d2e0539827b.png)

And to rephrase: This is because every location updates leads to two independent MQTT messages on two different MQTT topics, being processed at different time by my MQTT broker.

On a more philosophical level, **I know that 100% of my location points on home Assistant are incorrect.** 😂  
If the latitude just got updated, I have an outdated longitude. If the longitude just got updated, I have an outdated latitude.
_This is why you can see the location points jumping from "too much to the right" to "to much to the left" of the road... but never on it._

**Idea / Solution**
**Create a new MQTT topic teslamate/cars/$car_id/location containing two JSON attributes**
- Latitude 
- Longitude

So that every MQTT consumer can process both dimension of the location at once.

Cheers !
JLo

**Comment by binou06** [comment]
Hello,

I agree with you. In my case, got automation like opening the gate when approaching, lighting outdoor according to darkness outside. When parking the car (in reverse in the driveway), disarming alarm, when 1km from house, launching parfume diffuser for few minutes)... even done an interface to display distance between my modelY and my wife's (and display it on a map I can load on tesla browser). So when I know we will be on the same road (not the same direction) it allows us to have an idea of when we will come across to open the windows and make a sign 😉

By the way, I do not use HA. I do hand made (because I started years ago and even if results is nice, I like to "do things"). And I faced the same problem. 

So, here is how I made my "trick". Both dimensions (lat and lng) are broadcasted in a short delay. So, I use memcached ... 

- When my listener receive lat or lng, I check if the other dimension is stored in memcached. If not (that says that it is the first of the two values), I store the value in memcaced with a short timeout (few seconds) 
- If yes, I take the given value and the other dimension of coordinate (in memcached). To be clean, I devalidate memcached stored value. 
 
As a consequence, I only use these values as couple (full coordinates) and I do not have stairs effect like you as coordinate do not change one dimension by one dimension.

This is a "trick" and should be better to get a "full coordinate subscription" on MQTT... but it does the job for me. Do not know how HA is modular but it can be an idea.

Cheers

**Reply by jlpouffier** [reply]
I admire your dedication ☺️!

I think I can pull something similar in Home assistant.
However, I still believe a projet such as Teslamate should provide this out of the box for all consumers to use simply.

**Comment by binou06** [comment]
For sure +1 for the out of the box,

I do not know anything about elixir but after having a look at the code, I think it would be somewhere in `/teslamate/vehicles/vehicle.ex`. There seems to be lists defined that are enumerated in the MQTTpublisher lib. 

It seems that all received events from teslaAPI are "published" to MQTT... with just a layer on given "keys" to map  values (for model name correspondances for instance). Once again, I do not know anything about Elixir and did not spend enough time to understand properly... so excuse me if I say bullshits. Even if I'd like to contribute, I'm not legit to code anything with my poor (null) knowledge on that language.

I think that here could be defined `lat_lng` with the short term memory trick to join both dimensions... 

Adding memcached is IMHO overkill (because another "server listening" and source of goofs for "non geek people"). A simplistic algorithm like that should do the job :
**event : receiving `lat`**
- is `tmp_lng` empty ? (the other dimension)
-- yes : store `lat` value in `tmp_lat` for later use
-- no : publish `lat_lng` (object construct with `tmp_lat` and `tmp_lng`) and empty `tmp_lat` and `tmp_lng` variables

**event : receiving `lng`**
- is `tmp_lat` empty ? (the other dimension)
-- yes : store `lng` value in `tmp_lng` for later use
-- no : publish `lat_lng` (object construct witch `tmp_lat` and `tmp_lng`) and empty `tmp_lat` and `tmp_lng`

Problem : you could be shifted according to the order you receive events. On your exemple (with stairs effect) that means that this solution could lead to the line linking all "up corner" of the stair or "down corner" of the stair according to what dimension is received first (but more "reliable")

To be more accurate, could mimic the timeout I manage using memcached by storing "last dimension received timestamp" (either lat or lng) in, let say, `ts_last_received_dimension`... and when I say "is `tmp_xxx`empty" (in my idea upward)... just add AND ts_last_received_dimension not older than x seconds (x to be defined)... so that it will only "join" dimensions received in a short delay ( < x seconds), which means, "real" couple of dimensions.

Elixir dev nearby ???? 😜

**Comment by jlpouffier** [comment]
Hello 👋

In case someone see this issue and needs a bit of guidance on how to solve this on Home Assistant side,

[Here is my take on solving this issue](https://github.com/jlpouffier/home-assistant-config/blob/master/appdaemon/apps/tesla_automations.py)

It's a Python app running on AppDeamon.

It's pretty simple, it listens to latitude and longitude updates, stores the updates internally and sends them to another MQTT topic only if both updates have been received in a very short time frame.

Works very well.

Here is a picture of the results to be compared with the picture of the issue on the initial post.
![Screenshot_20230404-225254.png](https://user-images.githubusercontent.com/5878296/229918287-c7d63aef-8c3c-4fc3-9e04-64e3b58dc456.png)


I still believe this should be a feature of Tesla late out of the box...
But at least I managed to fix it on my side ;)
